collectors/htcondor/collector.py

```diff
@@ -0,0 +1,242 @@
+        assert type(cluster) == int and type(proc) == int, "Invalid job id"
+        self.logger.debug(
+            f'Querying HTCondor history for "{schedd_name}" starting from job "{cluster}.{proc}".'
+        )
+        cmd = [
+            "condor_history",
+            "-backwards",
+            "-name",
+            schedd_name,
+            "-since",
+            f"{cluster}.{proc}",
+            "-af:,",
+            *self.config.class_ads,
+        ]
+        if getattr(self.config, "pool", None):
+            cmd.extend(["-pool", self.config.pool])
+        if getattr(self.config, "job_status", None):
+            cmd.extend(
+                [
+                    "-constraint",
+                    f"\"{' || '.join(f'JobStatus == {j}' for j in self.config.job_status)}\"",
+                ]
+            )
+
+        self.logger.debug(f"Running command: \"{' '.join(cmd)}\"")
+
+        p = Popen(" ".join(cmd), stdout=PIPE, stderr=PIPE, shell=True)
+        output, err = p.communicate()
+        if err:
+            self.logger.error(f"Error querying HTCondor history:\n{err}")
+
+        jobs = [
+            map(maybe_convert, job.decode("utf-8").split(", "))
+            for job in output.strip().splitlines()
+        ]
+
+        return [dict(zip(self.config.class_ads, job)) for job in jobs]
+
+    def _generate_components(self, job: dict) -> List[Component]:
+        components = []
+        for component in self.config.components:
+            if amount := get_value(component, job):
+                try:
+                    # AUDITOR expects int-values for components
+                    amount = int(amount)
+                except ValueError:
+                    self.logger.warning(
+                        f"Could not convert amount ({amount}) for component "
+                        f"\"{component['name']}\" of job \"{job['GlobalJobId']}\" to int. "
+                        f"Skipping record."
+                    )
+                    raise ValueError
+                comp = Component(name=component["name"], amount=amount)
+                for score in component.get("scores", []):
+                    value = get_value(score, job) or "0.0"
+                    comp.with_score(Score(score["name"], maybe_convert(value)))
+                components.append(comp)
+        return components
+
+    def _get_meta(self, job: dict) -> Meta:
+        meta = Meta()
+        for key, entry in self.config.meta.items():
+            values = []
+            for item in entry if isinstance(entry, list) else [entry]:
+                if (value := get_value(item, job)) is not None:
+                    values.append(value)
+                    if key == "site":  # site is a special case
+                        break
+            if not values:
+                self.logger.warning(
+                    f"Could not find meta value for \"{key}\" for job \"{job['GlobalJobId']}\"."
```

Comment:
Here you escape `"`, in other log messages you start the f-string with `'`, maybe try do this in a consistent style.

Reply:
Consider to use raw formatting as `... for {key!r} for ...` (no quotes, but `!r` instead). This will automatically add quotes but also escape special characters that would otherwise be invisible.

Reply:
You are right, I've changed that to use raw strings for logging.